magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala

```diff
@@ -1408,61 +1440,136 @@ object DefaultRecordPersistence
     result
   }
 
+  private def opaQueriesToWhereParts(
+      tenantId: BigInt,
+      opaQueries: Seq[OpaQuery],
+      operationType: AuthOperations.OperationType
+  ): Iterable[SQLSyntax] = {
+    val theOpaQueries: List[OpaQuery] =
+      if (opaQueries.nonEmpty) opaQueries.toList
+      else List(OpaQuerySkipAccessControl)
+    val opaSql: SQLSyntax =
+      sqls"(${SQLSyntax
+          .joinWithOr(opaQueriesToWhereClauseParts(tenantId, theOpaQueries): _*)
+      })"
+
+    val result = List(opaSql)
+    result
+  }
+
+  private object ColumnNamePrefixType extends Enumeration {
+    type ColumnNamePrefixType
+    val PREFIX_TEMP: ColumnNamePrefixType.Value = Value(0, "temp")
+    val PREFIX_EMPTY: ColumnNamePrefixType.Value = Value(1, "")
+  }
+
+  private def getAspectColumnName(
+      index: Int,
+      prefix: ColumnNamePrefixType.Value = ColumnNamePrefixType.PREFIX_EMPTY
+  ): SQLSyntax = {
+    val prefixStr = if (prefix == ColumnNamePrefixType.PREFIX_TEMP) prefix.toString + "." else prefix.toString
+    // Use a simple numbered column name rather than trying to make the aspect name safe.
+    SQLSyntax.createUnsafely(prefixStr + s"aspect$index")
+  }
+
+  private def getOpaConditions(tenantId: BigInt, opaQueries: Seq[OpaQuery], operationType: AuthOperations.OperationType) = {
+    SQLSyntax.joinWithAnd(
+      opaQueriesToWhereParts(tenantId, opaQueries, operationType).toSeq: _*
+    )
+  }
+
   private def aspectIdsToSelectClauses(
+      session: DBSession,
       tenantId: BigInt,
       aspectIds: Iterable[String],
       operationType: AuthOperations.OperationType,
-      dereferenceDetails: Map[String, PropertyWithLink] = Map()
+      opaQueries: Seq[OpaQuery],
+      dereference: Boolean = false
   ): Iterable[scalikejdbc.SQLSyntax] = {
+    val opaConditions = getOpaConditions(tenantId, opaQueries, operationType)
+
     val result: Iterable[SQLSyntax] = aspectIds.zipWithIndex.map {
       case (aspectId, index) =>
-        // Use a simple numbered column name rather than trying to make the aspect name safe.
-        val aspectColumnName = SQLSyntax.createUnsafely(s"aspect$index")
-        val selection = dereferenceDetails
+        val referenceDetails =
```

Comment:
Could this be out of the loop (like on line 1490?)

Reply:
Yes. It was silly to be put in the loop.